crategen/validators.py

```diff
@@ -0,0 +1,109 @@
+from pydantic import ValidationError
+from typing import Union
+from .models import WRROCProcess, WRROCWorkflow, WRROCProvenance
+from urllib.parse import urlparse
+
+def validate_wrroc(data: dict) -> Union[WRROCProvenance, WRROCWorkflow, WRROCProcess]:
+    """
+    Validate that the input data is a valid WRROC entity and determine which profile it adheres to.
+    
+    This function attempts to validate the input data against the WRROCProvenance model first.
+    If that validation fails, it attempts validation against the WRROCWorkflow model.
+    If that also fails, it finally attempts validation against the WRROCProcess model.
+    
+    Args:
+        data (dict): The input data to validate.
+    
+    Returns:
+        Union[WRROCProvenance, WRROCWorkflow, WRROCProcess]: The validated WRROC data, indicating the highest profile the data adheres to.
+    
+    Raises:
+        ValueError: If the data does not adhere to any of the WRROC profiles.
+    """
+    # Convert '@id' to 'id' for validation purposes
+    if '@id' in data:
+        data['id'] = data.pop('@id')
+
+    errors = []
+
+    try:
+        return WRROCProvenance(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+
+    try:
+        return WRROCWorkflow(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+
+    try:
+        return WRROCProcess(**data)
+    except ValidationError as e:
+        errors.extend(e.errors())
+        raise ValueError(f"Invalid WRROC data: {errors}")
+
+def validate_wrroc_tes(data: dict) -> WRROCProcess:
+    """
+    Validate that the input data contains the fields required for WRROC to TES conversion.
+
+    This function first validates that the data is a valid WRROC entity by calling `validate_wrroc`.
+    Then it checks that the data contains all necessary fields for TES conversion.
+
+    Args:
+        data (dict): The input data to validate.
+
+    Returns:
+        WRROCProcess: The validated WRROC data that is suitable for TES conversion.
+
+    Raises:
+        ValueError: If the data is not valid WRROC data or does not contain the necessary fields for TES conversion.
+    """
+    validated_data = validate_wrroc(data)
+    required_fields = ["id", "name", "object", "result"]
```

Comment:
Couple of issues here:
1. Please use the `WRROCDataTES` model for the validation.
2. Assuming "object" and "result" are input and output files, I don't think these are required for TES. On the other hand, `executors` _is_ required, so we absolutely expect an "instrument" property. Please carefully check the TES specs and our [conversion tables](https://docs.google.com/spreadsheets/d/1JP9p2npUiQnt3a2hNeJqjfom8rcB7BDM4SV9HYcIu58/edit?gid=0#gid=0) and make sure the models are defined correctly.

Goes for WES as well, of course.

Reply:
In the end you'd have only this code in this validator (untested):

```py
data_validated = validate_wrroc(data=data)
try:
    data_wrroc_tes = WRROCDataTES(**data_validated)
except ValidationError as exc:
    raise ValueError(
        f"WRROC data insufficient for TES conversion: {exc.errors()}"
    ) from exc
return data_wrroc_tes